### PR TITLE
Create cdn_url method to handle image content direct access

### PR DESCRIPTION
**Pull request by pacodelaluna**
#### What? Why?

The objective is to find a way to render image attachments directly from the cloud storage solution (AWS S3).
We are facing some performance issues on FR Prod. It may not be the only reason, but to reduce Nginx and Puma stress can help improving performance a bit. Especially that these files are actually public.

Related to https://github.com/openfoodfoundation/openfoodnetwork/issues/9859.

#### What should we test?
The feature will be active on Admin side first.
- Visit any pages listing products on Admin side
- Visit a Image Edit page where a new image version is shown (CDN Image)

#### Release notes

Changelog Category: Technical changes

### CODE REVIEW

File: app/models/spree/image.rb [hunk 1]
```diff
@@ -35,6 +35,12 @@ def url(size)
       Rails.application.routes.url_helpers.url_for(variant(size))
     end
 
+    def cdn_url(size)
+      return url(size) if ENV["S3_BUCKET"].nil?
+
+      variant(size).service_url
```
Comment (pacodelaluna) [inline app/models/spree/image.rb:41]: It is not suggested to use this method publicly, but it should not be such an issue in our case as the images we want to display are actually public. I have not find an easier way to display the image directly from the cloud storage, this method is setting up all we need.

File: app/views/spree/admin/images/edit.html.haml [hunk 2]
```diff
@@ -14,6 +14,10 @@
       - # A Rails bug makes it necessary to call `main_app.url_for` here.
       - # https://github.com/rails/rails/issues/31325
       = link_to image_tag(main_app.url_for(@image.variant(:small))), main_app.url_for(@image.variant(:product))
+    .field.alpha.three.columns.align-center
+      = f.label t('spree.cdn_thumbnail')
```
Comment (pacodelaluna) [inline app/views/spree/admin/images/edit.html.haml:18]: Translations will be updated once the solution is validated.

**Review by mkllnk — COMMENTED** [review]
Worth investigating. This needs a bit of testing though. Several thoughts:

* `service_url` is deprecated and we should use the blob URL, I think: https://stackoverflow.com/questions/49791684/how-to-retrieve-attachment-url-with-rails-active-storage-with-s3
* When an image is uploaded, we don't generate all the variants straight away. So, the image may actually not be there when we publish the URL? Or, this method could be waiting for the image to be rescaled and uploaded which would make the performance a lot worse.
* We need to find a solution for the shopfront because that's where most of the requests come from. What about adding `loading="lazy"` to our image tags instead?
* The redirect is extremely quick compared to other actions in our app. If this is a problem, maybe we need to increase the connection pool size.

While I think that this is worth looking into, I don't have the capacity to do that right now and I'm not sure it's the best approach. There are probably good reasons why Rails decided to do it this way.

**Review by pacodelaluna — COMMENTED** [review]
@mkllnk Thanks for your review, I have added your suggestions. Trying to add some specs right now.

I understand your point about the redirect request being really fast, but do we really need to control the image url at Rails side ? ActiveStorage is also providing an helper to get the direct blob url, `rails_blob_path`, so the direct rendering seems to be something supported, but not by default, I don't know exactly why.

On FR Prod, `ActiveStorage::Representations::RedirectController#show` action is the third one in term of time spent on the resources.

**Comment by mkllnk** [comment]
> ActiveStorage is also providing an helper to get the direct blob url, rails_blob_path, so the direct rendering seems to be something supported, but not by default, I don't know exactly why.

I think that the problem with this is that the method has to wait until the image is resized and uploaded. So the first time a shopfront with new images is loaded the request would take ages. Is there a way to find out if the variant is generated already? Then we could take the processed variant if available and the redirect if it needs more time to process and upload.

> On FR Prod, ActiveStorage::Representations::RedirectController#show action is the third one in term of time spent on the resources.

That's a good point. I'm wondering if this was improved in Rails 7.

**Comment by mkllnk** [comment]
Here's a bit more info: https://api.rubyonrails.org/v6.1.4/classes/ActiveStorage/Variant.html

I just checked and you can call `processed?` on the variant. So our `Spree::Image#url` method could check that and then provide the process URL directly when it's available.

**Comment by sigmundpetersen** [comment]
Are we moving forward with this? Or should we close?
What do you think @pacodelaluna

**Comment by pacodelaluna** [comment]
@sigmundpetersen Sorry I was busy on other topics lately, I didn't communicate properly. I still want to give a try to this, just give me 2 more weeks and I will propose something based on the suggestions.

**Comment by sigmundpetersen** [comment]
Sure, no worries!
Take the time you need 🙂

**Comment by pacodelaluna** [comment]
If we mix the approaches, redirect urls and s3 urls, this means that we will need to wait for a ´HEAD´ request to S3 in order to know if the file is present, and this can create then a bottleneck on our server if we have large images list, especially if this information is not saved somewhere.


An other idea is to automatically process the variants after the image creation. It is still not straight-forward as we need to wait the ´ActiveStorage::AnalyzeJob´ job that will give some required information about the image (height, width, content type, ...). It can also add extra files (unrequested variants) on S3 Storage compared to present solution (on-the-fly generation).


Something surprising is that the analyse result is saved inside the ´ActiveStorage::Blob´ object metadata, but not the process result... I am not sure why but to handle this could give us the expected behavior. I will dig it.


```
irb(main):010:0> Spree::Image.last.variant(:small).processed
  S3 Storage (1926.6ms) Downloaded file from key: 3vg4v3c2quzs1wou8x34f20z6jlo
  S3 Storage (473.9ms) Uploaded file to key: qx2lum8t2hecppoe6ds092bp5b3b (checksum: TmY5h44xVjMMuEsvp5Es3Q==)
Enqueued ActiveStorage::AnalyzeJob (Job ID: 6fad51b7-bdcb-4c67-9128-a7544d9e13ac) to Sidekiq(default) with arguments: #<GlobalID:0x0000559417361680 @uri=#<URI::GID gid://openfoodnetwork/ActiveStorage::Blob/92278>>
=> 
#<ActiveStorage::VariantWithRecord:0x000055941bad00c0
 @blob=
  #<ActiveStorage::Blob:0x000055941bad11f0
   id: 92246,
   key: "3vg4v3c2quzs1wou8x34f20z6jlo",
   filename: "Origan feuille.jpg",
   content_type: "image/jpeg",
   metadata: {"identified"=>true, "width"=>3456, "height"=>3456, "analyzed"=>true},
   service_name: "amazon",
   byte_size: 2723092,
   checksum: "ubztL7Scd6/26B1piagGxg==",
   created_at: Wed, 04 Jan 2023 20:10:14.164399000 CET +01:00>,
 @record=#<ActiveStorage::VariantRecord:0x00005594136f25a0 id: 58497, blob_id: 92246, variation_digest: "H5v0r22GDnZpKfLoIx8yZVpMjLo=">,
 @variation=#<ActiveStorage::Variation:0x000055941bad0250 @transformations={:format=>"jpg", :resize_to_fill=>[227, 227]}>>
```

**Comment by mkllnk** [comment]
> need to wait for a HEAD request to S3 in order to know if the file is present

Rails can store the result in the database. The `ActiveStorage::VariantWithRecord` in your example is exactly that record. I haven't tested it but I would imagine that Rails is not sending a HEAD request every time.

A very low tech solution would be to show all variants when you upload an image. The upside is that it's a simple view referencing images and that you can verify that all sizes look good after uploading. Of course, it will take a while to load the page but you can dismiss that modal if you trust the system. The requests for the different variants have then been made by the browser.

**Comment by RachL** [comment]
@pacodelaluna just checking in: do you still plan to work on this PR or should we close it for now?

**Comment by pacodelaluna** [comment]
@RachL No time lately unfortunately. Also it became less important as the performance since better with other improvements. You can close then it, I will reopen it when/if I foind time to work on @mkllnk 's suggestions.

**Pull request closed without merging**